tcn_hpl/data/utils/pose_generation/generate_pose_data.py

```diff
@@ -167,195 +229,267 @@ def predict_single(self, image: torch.tensor) -> list:
-        - It's assumed that `self.dataset` is already loaded or initialized and is an instance compatible with the
-        processing expected by `generate_bbs_and_pose`.
-        """
-        self.dataset = self.generate_bbs_and_pose(self.dataset)
-
-        dataset_path_with_pose = f"{self.config[self.config_data_key]['save_root']}/{self.dataset_path_name}_with_dets_and_pose.mscoco.json"
-        self.dataset.dump(dataset_path_with_pose, newlines=True)
-        print(f"Saved test dataset to: {dataset_path_with_pose}")
-        return
-
-
-def main():
-
-    main_config_path = f"configs/main.yaml"
-    config = load_yaml_as_dict(main_config_path)
-
-    PG = PosesGenerator(config)
-    PG.run()
+        for img_id in tqdm(
+            out_dset.images(),
+            desc="Processing images",
+            unit="images",
+        ):
+            img_path = Path(out_dset.get_image_fpath(img_id)).as_posix()
+            img = read_image(img_path, format="BGR")
+            boxes, scores, classes, keypoints_list = self.predict_single(img)
+
+            # We will need non-numpy data types to insert into the structure to
+            # follow JSON compliance.
+            boxes_list = boxes.tolist()
+            scores_list = scores.tolist()
+            classes_list = classes.tolist()
+
+            # Construct annotations for predictions.
+            for box, score, cls_idx, kp_mat in zip(
+                boxes_list, scores_list, classes_list, keypoints_list
+            ):
+                # Convert keypoints from scored XY coordinates to the COCO
+                # notation with visibility.
+                kp_kw = {}
+                if kp_mat is not None:
+                    kp_vals = []
+                    # According to spec (https://cocodataset.org/#format-data):
+                    # visibility flag v defined as v=0: not labeled (in
+                    # which case x=y=0), v=1: labeled but not visible, and
+                    # v=2: labeled and visible.
+                    for kp in kp_mat.tolist():
+                        # TODO: Filter keypoints if present by a threshold?
+                        #       if not above threshold, fill in triple-0.
+                        kp_vals.extend([*kp[:2], 2])
+                    kp_kw["keypoints"] = kp_vals
+                    kp_kw["keypoint_scores"] = kp_mat[:, 2].ravel().tolist()
+
+                out_dset.add_annotation(
+                    image_id=img_id,
+                    category_id=cls_idx,
+                    bbox=box,
+                    score=score,
+                    **kp_kw,
+                )
+
+            # Checkpoint output the COCO Dataset if we are at some interval
+            if on_image_done_callback is not None:
+                on_image_done_callback(out_dset)
+
+        return out_dset
+
+
+@click.command(context_settings={"help_option_names": ["-h", "--help"]})
```

Comment:
I think another @click.option that might make sense to add will be joint-wise confidence threshold? But we can hold off on that for now.

Reply:
I have a TODO in there for that. I can add that relatively quickly.